### ISSUE TITLE
Crash on freeing a node in `_exit_tree()` with `free()`

### ISSUE THREAD

**Issue by ObaniGemini** [godotengine/godot#63742]
### Godot version

3.4.4.stable.arch_linux

### System information

Arch Linux, GLES 3.0, NVIDIA GeForce GTX 1050 Ti/PCIe/SSE2

### Issue description

While using **remove_child** to keep a node's instance available, using `free()` in the node's `_exit_tree()` function makes it crash, even while avoiding a stack overflow loop.

I need to use `free()` because **remove_child** removes the child from the tree without deleting it, but I sometimes need to delete it when it exits the tree, and `queue_free()` only works if the node is in the scene tree.

### Steps to reproduce

Have a parent node **P** with a child **C**, both with scripts.

In **P**'s script, keep a variable that contains **C**'s reference, and call remove_child(**C**)
```
onready var child = $child

func _ready():
	remove_child(child)
```

In **C**'s script, in _exit_tree(), free the instance : 
```
var deleting = false

func _exit_tree():
	if !deleting and !is_queued_for_deletion():
		deleting = true #necessary to not have a stack overflow
		free()
```

### Minimal reproduction project

[BugTest.zip](https://github.com/godotengine/godot/files/9229705/BugTest.zip)

**Comment by AnidemDex** [comment]
Using
```gdscript
extends Node

var clone
var original = true

func _notification(what: int) -> void:
	match what:
		NOTIFICATION_ENTER_TREE:
			prints(self, "NOTIFICATION_ENTER_TREE")
		
		NOTIFICATION_READY:
			prints(self, "NOTIFICATION_READY")
			if original:
				clone = self.duplicate(4)
				clone.original = false
				clone.name = "Clone"
				add_child(clone)
				yield(get_tree().create_timer(1), "timeout")
				remove_child(clone)
			
		NOTIFICATION_EXIT_TREE:
		        free()
		        #queue_free()
			prints(self, "NOTIFICATION_EXIT_TREE")
		
		NOTIFICATION_PREDELETE:
			prints(self, "NOTIFICATION_PREDELETE")
```
An stack overflow is generated from `clone`:
![image](https://user-images.githubusercontent.com/7025991/182045660-bae2eb52-a73d-49be-9800-65bac30bc51a.png)

Anyway, you can use `queue_free()` in `NOTIFICATION_EXIT_TREE` (`_exit_tree()`) without problems:
![image](https://user-images.githubusercontent.com/7025991/182045756-02b6c648-728a-4116-9957-f60bd9e7fa69.png)

Using v3.4.2.stable.official [45eaa2daf]

**Comment by ObaniGemini** [comment]
No, the stack overflow happends because free makes you enter into the _exit_tree state.
That's why I added a bool called `deleting` in that case. Also, queue_free() doesn't work for me if you're already exitting the tree.

**Comment by timothyqiu** [comment]
I think you can use something like this:

```gdscript
func _exit_tree():
    call_deferred("free")
```

* `_exit_tree()` is called when the node is _about to_ exit tree. So calling `free()` directly will make the node to exit tree again.
* There're still some internal logic after calling back `_exit_tree()` that uses the node object, so you can't free the node there.

**Comment by akien-mga** [comment]
Yeah I mentioned on chat that it may not work outside the scene tree but I actually wrote the code above in 2017 so it works :P

**Comment by YuriSizov** [comment]
The solution should probably be erring if the node is currently exiting, or locked for any other reason for that matter, if you call `free` on it.

**Comment by saierXP** [comment]
I cannot reproduce after running 30 times since godot 4.6.1. In 4.5-stable runtime window will crash once in several running the scene.
